lib/vector-core/src/enrichment_table/enrichment_tables.rs

```diff
@@ -0,0 +1,286 @@
+use super::EnrichmentTable;
+use arc_swap::ArcSwap;
+#[cfg(feature = "vrl")]
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::{collections::HashMap, sync::Mutex};
+
+/// The Enrichment Tables manages the collection of `EnrichmentTable`s loaded into Vector.
+/// Enrichment Tables go through two stages.
+///
+/// ## 1. Writing
+///
+/// The  tables are loaded. There are two elements that need loading. The first is the actual data.
+/// This is loaded at config load time, the actual loading is performed by the implementation of
+/// the `EnrichmentTable` trait. Next, the tables are passed through Vectors `Transform` components,
+/// particularly the `Remap` transform. These Transforms are able to determine which fields we will
+/// want to lookup whilst Vector is running. They can notify the tables of these fields no that the
+/// data can be indexed.
+///
+/// During this phase, the data needs to be mutated and shared around a number of potential
+/// different components. Also, performance is not absolutely critical at this phase. Thus, the
+/// data is stored within a `Mutex` - stored in the `loading` field of this struct.
+///
+/// ## 2. Reading
+///
+/// Once all the data has been loaded we can move to the next stage. This is signified by calling
+/// the `finish_load` method. At this point all the data is swapped out of the `Mutex` in the loading
+/// field into the `ArcSwap` of the `tables` field. `ArcSwap` provides lock-free read-only access
+/// to the data. From this point on we have fast, efficient read-only access and can no longer add
+/// indexes or otherwise mutate the data .
+///
+/// Cloning this object is designed to be cheap. The underlying data will be shared by all clones.
+///
+#[derive(Clone, Default)]
+pub struct EnrichmentTables {
+    loading: Arc<Mutex<Option<HashMap<String, Box<dyn EnrichmentTable + Send + Sync>>>>>,
+    tables: Arc<ArcSwap<Option<HashMap<String, Box<dyn EnrichmentTable + Send + Sync>>>>>,
```

Comment:
Just a random thought, right now, the loading vs available state is at the _collection level_, does it make sense to move this down into the actual individual enrichment tables? I.e. tables are loaded concurrently, and some can be ready while others aren't yet?

I wonder if this would help in the future (the "in the future" part might be a reason not to do this right now, unless it's hard to change this later) when you have multiple pipelines using multiple different enrichment tables where some pipelines can start processing data within seconds, while others take minutes to load their respective enrichment tables.

Reply:
The way `ArcSwap` works is pretty all or nothing. You only get to update (swap in) the entire collection. It does give you lock free read only access though, which is what this use case is optimized for. If we were to move the ArcSwap down to the table level, we would need to wrap that collection in a separate type of lock, which defeats the purpose.

This does mean that when we update a table should the underlying data change, we will have to update the entire collection - either by cloning the unchanged data, or forcing them to reload anyway. 

There are a lot of compromised made for writing the data with these tables, but we are optimizing for safe and fast reads which is the primary use case.